gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -252,24 +255,52 @@ private WriteObjectResponse doResumableUpload() throws IOException {
 
       boolean objectFinalized = false;
       while (!objectFinalized) {
-        WriteObjectRequest insertRequest;
+        WriteObjectRequest insertRequest = null;
         if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
           insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
         } else {
-          ByteString data =
-              ByteString.readFrom(
-                  ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
-          dataChunkMap.put(writeOffset, data);
           if (dataChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-            dataChunkMap.remove(dataChunkMap.firstKey());
+            /*
+             If dataChunkMap is full, get the committedWriteOffset. This will make a API call to the
+             server and add latency in this path/context. Since there are already chunks in flight,
+             calling this API will not reduce overall throughput. It will throttle the upstream
+             write call now rather than onFinalize, which is fine. This will also increase QPS to
+             the GCS backend. The increase will be linear to number of chunks written, so that
+             should also be fine.
+            */
+            committedWriteOffset = getCommittedWriteSizeWithRetries(uploadId);
```

Comment:
Can this be done asynchronously ?

Reply:
It could be done, but then we will need a block/signal code to get back control. Moreover, when we are calling this API, we (this thread) are idle, as we are not able to write subsequent buffers. So for simplicity sake, I would prefer this approach.